### PR TITLE
Fix nested non-byref like VC with no pointer fields

### LINKED ISSUE #80624: .NET 7 JIT Crash when referencing struct of particular shape: Fatal error. Internal CLR error. (0x80131506)
### Description

Referencing a ref struct with explicit layout, containing a ref struct, containing a struct, containing a value crashes the JIT in the importer phase when jitting the method containing that reference.

### Reproduction Steps

```csharp
using System.Runtime.InteropServices;

Console.WriteLine("Testing for crash.");
FailsToJit();
Console.WriteLine("Did not crash.");

static void FailsToJit()
{
	// any reference (direct or indirect) to godot_variant will crash in during JIT import when resolving the type token
	// loading a function pointer using the type, sizeof(), ... all crash in basically the same way
	void Test(godot_variant arg) { }
	Test(default);
}

[StructLayout(LayoutKind.Explicit)]
public ref struct godot_variant
{
	[FieldOffset(0)] // offset doesn't matter
	private godot_variant_data _data;

	private ref struct godot_variant_data
	{
		private godot_variant_data_mem _mem;

		public struct godot_variant_data_mem
		{
			private float _mem0; // also happens for int here
		}
	}
}
```

### Expected behavior

Program runs to completion, printing two lines to console.

### Actual behavior

The program crashes with the following output when trying to execute `FailsToJit`.
```
Fatal error. Internal CLR error. (0x80131506)
   at Program.<<Main>$>g__FailsToJit|0_0()
   at Program.<Main>$(System.String[])
```

Native Backtrace:
```
>	clrjit.dll!Compiler::lvaInitTypeRef() Line 280	C++
 	clrjit.dll!Compiler::compCompileHelper(CORINFO_MODULE_STRUCT_ * classPtr, ICorJitInfo * compHnd, CORINFO_METHOD_INFO * methodInfo, void * * methodCodePtr, unsigned int * methodCodeSize, JitFlags * compileFlags) Line 6509	C++
 	clrjit.dll!`Compiler::compCompile'[::M]::__Body::Run(Compiler::compCompile::__l2::__JITParam * __JITpParam) Line 5930	C++
 	clrjit.dll!Compiler::compCompile(CORINFO_MODULE_STRUCT_ * classPtr, void * * methodCodePtr, unsigned int * methodCodeSize, JitFlags * compileFlags) Line 5934	C++
 	clrjit.dll!``jitNativeCode'::`8'::__Body::Run'::`6'::__Body::Run(jitNativeCode::__l8::__Body::Run::__l5::__JITParam * __JITpParam) Line 7413	C++
 	clrjit.dll!`jitNativeCode'::`8'::__Body::Run(jitNativeCode::__l2::__JITParam * __JITpParam) Line 7416	C++
 	clrjit.dll!jitNativeCode(CORINFO_METHOD_STRUCT_ * methodHnd, CORINFO_MODULE_STRUCT_ * classPtr, ICorJitInfo * compHnd, CORINFO_METHOD_INFO * methodInfo, void * * methodCodePtr, unsigned int * methodCodeSize, JitFlags * compileFlags, void * inlineInfoPtr) Line 7440	C++
 	clrjit.dll!CILJit::compileMethod(ICorJitInfo * compHnd, CORINFO_METHOD_INFO * methodInfo, unsigned int flags, unsigned char * * entryAddress, unsigned int * nativeSizeOfCode) Line 261	C++
 	[Managed to Native Transition]	
 	test.dll!Program.<Main>$(string[] args) Line 5	C#
 	[Native to Managed Transition]	
 	[Inline Frame] hostpolicy.dll!coreclr_t::execute_assembly(int) Line 89	C++
 	hostpolicy.dll!run_app_for_context(const hostpolicy_context_t & context, int argc, const wchar_t * * argv) Line 255	C++
 	hostpolicy.dll!run_app(const int argc, const wchar_t * * argv) Line 284	C++
 	hostpolicy.dll!corehost_main(const int argc, const wchar_t * * argv) Line 430	C++
 	hostfxr.dll!execute_app(const std::wstring & impl_dll_dir, corehost_init_t * init, const int argc, const wchar_t * * argv) Line 146	C++
 	hostfxr.dll!`anonymous namespace'::read_config_and_execute(const std::wstring & host_command, const host_startup_info_t & host_info, const std::wstring & app_candidate, const std::unordered_map<enum known_options,std::vector<std::wstring,std::allocator<std::wstring>>,known_options_hash,std::equal_to<enum known_options>,std::allocator<std::pair<enum known_options const ,std::vector<std::wstring,std::allocator<std::wstring>>>>> & opts, int new_argc, const wchar_t * * new_argv, host_mode_t mode, const bool is_sdk_command, wchar_t * out_buffer, int buffer_size, int * required_buffer_size) Line 533	C++
 	hostfxr.dll!fx_muxer_t::handle_exec_host_command(const std::wstring & host_command, const host_startup_info_t & host_info, const std::wstring & app_candidate, const std::unordered_map<enum known_options,std::vector<std::wstring,std::allocator<std::wstring>>,known_options_hash,std::equal_to<enum known_options>,std::allocator<std::pair<enum known_options const ,std::vector<std::wstring,std::allocator<std::wstring>>>>> & opts, int argc, const wchar_t * * argv, int argoff, host_mode_t mode, const bool is_sdk_command, wchar_t * result_buffer, int buffer_size, int * required_buffer_size) Line 1018	C++
 	hostfxr.dll!fx_muxer_t::execute(const std::wstring host_command, const int argc, const wchar_t * * argv, const host_startup_info_t & host_info, wchar_t * result_buffer, int buffer_size, int * required_buffer_size) Line 579	C++
 	hostfxr.dll!hostfxr_main_startupinfo(const int argc, const wchar_t * * argv, const wchar_t * host_path, const wchar_t * dotnet_root, const wchar_t * app_path) Line 61	C++
 	test.exe!exe_start(const int argc, const wchar_t * * argv) Line 251	C++
 	test.exe!wmain(const int argc, const wchar_t * * argv) Line 322	C++
 	[Inline Frame] test.exe!invoke_main() Line 90	C++
 	test.exe!__scrt_common_main_seh() Line 288	C++
 	kernel32.dll!BaseThreadInitThunk()	Unknown
 	ntdll.dll!RtlUserThreadStart()	Unknown
```

Native debugging message:
```
Exception thrown: read access violation.
pSeries was 0x7FF7DB2478D8.
```

It varies depending on how the type is used, but it always ends at a call `info.compCompHnd->xxx()` with a type token as one of the arguments.
I think execution should continue in `coreclr.dll` at this point, but i am unable to get visual studio to step into that call, or to execute the program with a debug build of coreclr

### Regression?

Yes, when changing the target framework to net6.0 the same code does not crash and executes successfully.

### Known Workarounds

Removing about anything from the MRP makes the crash go away.

### Configuration

On Windows 10 x64 versions:
.NET 7.0.2 - d099f075e45d2aa6007a22b71b45a08758559f80
.NET 7.0.0 - d037e070ebe5c83838443f869d5800752b0fcb13

On Linux Ubuntu 22.04.1 x64 version
.NET 7.0.2 - d037e070ebe5c83838443f869d5800752b0fcb13

### Other information

_No response_

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Tests need to be added. This is just for checking existing tests on all platforms first.

Fixes https://github.com/dotnet/runtime/issues/80624

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
@BrzVlad @vargaz @lambdageek Any thoughts on the llvmfullaot legs?

**Comment by vargaz** [comment]
The failure is in the newly added tests, so it definitely looks like a mono problem. I'd suggest disabling it for now for mono.

File: src/tests/Loader/classloader/explicitlayout/objrefandnonobjrefoverlap/case13.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        {
+            return typeof(Explicit1).ToString();
+        }
+    }
+
+    [Fact]
+    public static void Validate_Explicit2()
+    {
+        Load2();
+        // that's enough. if we didn't throw a TypeLoadException, the test case will fail.
```
Comment (AaronRobinsonMSFT) [inline src/tests/Loader/classloader/explicitlayout/objrefandnonobjrefoverlap/case13.cs:74]:
```suggestion
```

File: src/tests/Loader/classloader/explicitlayout/objrefandnonobjrefoverlap/case13.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        {
+            return typeof(Explicit2).ToString();
+        }
+    }
+
+    [Fact]
+    public static void Validate_Explicit3()
+    {
+        Load3();
+        // that's enough. if we didn't throw a TypeLoadException, the test case will fail.
```
Comment (AaronRobinsonMSFT) [inline src/tests/Loader/classloader/explicitlayout/objrefandnonobjrefoverlap/case13.cs:87]:
```suggestion
```

File: src/tests/Loader/classloader/explicitlayout/objrefandnonobjrefoverlap/case13.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        {
+            public WithByRefs Field2;
+        }
+    }
+
+    [Fact]
+    public static void Validate_Explicit1()
+    {
+        Load1();
+        // that's enough. if we didn't throw a TypeLoadException, the test case will fail.
```
Comment (AaronRobinsonMSFT) [inline src/tests/Loader/classloader/explicitlayout/objrefandnonobjrefoverlap/case13.cs:61]:
```suggestion
```

**Comment by lambdageek** [comment]
The issue in mono (I commented in detail on #57350) is probably that we're trying to throw a TLE during AOT compilation.
Instead we should just print out "can't compile [assembly] due to: [whatever the problem is with the test class]".  But fundamentally that will still leave the new test to be skipped by the Mono AOT lanes - the runtime test suite doesn't consider "AOT compiler failed on this input assembly" a pass.

I'm not sure we can defer the error until execution time - what if the bad class was used in a method signature - the method wouldn't even be callable.

I'm curious what NativeAOT does with such input assemblies

File: src/tests/issues.targets [hunk 4]
```diff
@@ -2988,10 +2988,13 @@
...
         </ExcludeList>
-        <ExcludeList Include = "$(XunitTestBinBase)/baseservices/TieredCompilation/TieredVtableMethodTests/**">
+        <ExcludeList Include="$(XunitTestBinBase)/baseservices/TieredCompilation/TieredVtableMethodTests/**">
+            <Issue>https://github.com/dotnet/runtime/issues/57350</Issue>
+        </ExcludeList>
+        <ExcludeList Include="$(XunitTestBinBase)/Loader/classloader/explicitlayout/objrefandnonobjrefoverlap/case2/**">
             <Issue>https://github.com/dotnet/runtime/issues/57350</Issue>
         </ExcludeList>
-        <ExcludeList Include = "$(XunitTestBinBase)/Loader/classloader/explicitlayout/objrefandnonobjrefoverlap/case2/**">
+        <ExcludeList Include="$(XunitTestBinBase)/Loader/classloader/explicitlayout/objrefandnonobjrefoverlap/case13/**">
```
Comment (lambdageek) [inline src/tests/issues.targets:2997]: I'm surprised this is necessary: Directory.Build.props for the whole `objrefandnonobjrefoverlap` directory specifies `<MonoAotIncompatible>true</MonoAotIncompatible>` which should skip the test.

That said, I think #57350 is masking a real issue here; arguably `MonoAotIncompatible` is wrong here anyway - the test should AOT and pass.

Comment (lambdageek) [reply src/tests/issues.targets:2997]: I see: I misunderstood the intent of `MonoAotIncompatible`.  It only blocks AOT compilation.  But it doesn't prevent the test from attempting to run. Hence we need to skip it here in the FullAOT issues.targets section.

Comment (AaronRobinsonMSFT) [reply src/tests/issues.targets:2997]: @jkoritzinsky Just to confirm the scenario above. The MonoAOT scenarios require two blockers, right? The first is for the AOT step and the second, in issues.targets, is for running that particular test.

Comment (lambdageek) [reply src/tests/issues.targets:2997]: Also maybe the right thing for case13 (which is a positive test) is to **un**set `MonoAotIncompatible` overriding what's in Directories.Build.props; and not to add to issues.targets.  This test should pass.  Evidently it passes on the Mono non-AOT scenarios - so we don't throw an incorrect loader exception or anything like that.  It's just that we skipped AOTing it, and now the AOT-only runtime is complaining that it's being asked to JIT a test case.

Comment (jkoritzinsky) [reply src/tests/issues.targets:2997]: Yes, you need to add two blocks. The MonoAotIncompatible blocks the AOT step and the issues.targets blocks the execution.

Comment (AaronRobinsonMSFT) [reply src/tests/issues.targets:2997]: @lambdageek Sure. I can give that a shot.

**Comment by AaronRobinsonMSFT** [comment]
@mangod9 Can you suggest someone who could help review this change? @davidwrighton is away.

**Comment by mangod9** [comment]
@trylek, who can take a look.

**Review by trylek — APPROVED** [review]
LGTM, thanks for fixing this!

**Pull request merged**